### ISSUE TITLE
Sign in success callback not working for phone numbers

### ISSUE THREAD

**Issue by aswerdlow935** [firebase/firebaseui-web#184]
For phone number verificaiton, my code for email and google accounts doesn't seem to be running at all using this function (signInSuccess: function(currentUser, credential, redirectUrl) {}). Is this intentional or a bug?

**Comment by bojeil-google** [comment]
Hey @aswerdlow935, do you happen to use ReactJS? there have been reports of this which I haven't been able to reproduce:
https://github.com/firebase/firebaseui-web/issues/149

signInSuccess callback should always run on successful sign-in. If you can provide a way for me to reliably replicate this, it would be very helpful.

**Comment by aswerdlow935** [comment]
Here is the Github gist: https://gist.github.com/aswerdlow935/f2571f82a912d5393145e73736f95329
Both email and google sign in gave me an alert saying 'firebase' but using a phone number didn't give me an alert. I am trying to add the users to my realtime database at account creation client-side  which is why I call my addUser function, but the alert comes before that and does not triggers when using a phone number. Let me know if there's any other information that might be helpful :)

**Comment by bojeil-google** [comment]
Ok, I just tried your code with phone number sign-in. signInSuccess triggers as expected. I am getting the alert messages:
Ran Sign in Success
Check your email to verify your account
In general you are not chaining your promises.

First off, a phone number user doesn't have an email so you can't send an email verification as there is no email to verify.
Second, addUser will run in paralllel as signInSuccess. It will likely be interrupted and not allowed to complete in time.
Change addUser to return the db promise to set the user data. Change your callback to the following:

```
        signInSuccess: function(currentUser, credential, redirectUrl) {
          alert('Ran Sign in Success');
          addUser().then(function() {
            if (firebase.auth().currentUser.email && !firebase.auth().currentUser.emailVerified) {
              return firebase.auth().currentUser.sendEmailVerification();
            }
          }).then(function() {
            // All stuff done.
            window.location.href = 'loggedIn.html';
          }).catch(function(error) {
            alert(JSON.stringify(error));
          });
          // Do not redirect.
          return false;
        }

    function addUser() {
      var user = firebase.auth().currentUser;
      return db.ref('users/' + user.uid).set({
        provider: user.providerData,
        name: user.displayName,
        email: user.email
      });
    }
```

**Comment by aswerdlow935** [comment]
That's really weird. You are right about needing to chain my calls, and your method of checking if the account has an email address works very well, but when I run the original code in the gist without changing anything but adding the project config, I do not get any alert messages with phone numbers. I get the two alerts for email and google accounts but still do not get it for phone numbers.  Could this be a project config thing? The phone number accounts show up in the auth section in the firebase console, and I the code I ran is the exact same I showed. The phone number login just continues to the loggedIn page without any alerts, which is why to me it looks like signIn success is not running.

**Comment by bojeil-google** [comment]
I tried it yesterday and it was working for phone numbers. Please apply my fix. Your original code is incorrect.

**Comment by aswerdlow935** [comment]
I believe I figured out the issue. I was calling firebase.auth().onAuthStateChanged once the page loaded in order to check if the user was logged in and if so, redirect them to the appropriate page. From my testing, it seem that this function was redirecting to the appropriate page once user was true, which mean the second the account was created and the function ran, it interrupted the email verification and account creation. It's weird why originally it only happened with phone number login, but it started preventing email verification also, so I simply added a boolean flag that was set to false at the beginning of signInSuccess that stopped onAuthStateChanged from redirecting, and let my code in signInSuccess handle that. I used your code and made this fix and now everything works. Maybe what I was doing is uncommon, but if anyone else has this problem, I hope this helps.
